File: easy/strings/python/8-atoi.py
from curses.ascii import isdigit


class Solution:
    def myAtoi(self, s: str) -> int:
        s = s.strip()

        if len(s) == 0:
            return 0

        sign = "+"
        if s[0] == "-" or s[0] == "+":
            sign = s[0]
            s = s[1:]

        if len(s) == 0:
            return 0

        if not s[0].isdigit():
            return 0

        chars = []
        for c in s:
            if not c.isdigit():
                break
            chars.append(c)

        num = int("".join(d for d in chars))
        num = num * -1 if sign == "-" else num
        if -2**31 <= num <= 2**31-1:
            return num

        if num < -2**31:
            return -2**31

        if num > 2**31 - 1:
            return 2**31 - 1

        return 0


# print(Solution().myAtoi("42"))
# print(Solution().myAtoi("      -42"))
# print(Solution().myAtoi("4193 with words"))
# print(Solution().myAtoi("words with 123"))
print(Solution().myAtoi("+1"))

File: easy/strings/cpp/8-atoi.cpp
#include "../../../stdc++.h"

using std::string;
using std::vector;

class Solution
{
public:
    int myAtoi(string s)
    {
        int start = 0;
        bool sign = true;
        for (size_t i = 0; i < s.size(); i++)
        {
            if (s[i] != ' ')
            {
                start = i;
                if (s[i] == '-' || s[i] == '+')
                {
                    sign = s[i] == '+';
                    start = i + 1;
                }
                break;
            }
        }
        string::iterator it = s.begin() + start;
        long ans = 0;
        while (isdigit(*it) && it != s.end())
        {
            ans = ans * 10 + (*it - '0');
            if (ans > INT_MAX && sign)
            {
                return INT_MAX;
            }
            if (ans > INT_MAX && !sign)
            {
                return INT_MIN;
            }
            it++;
        }
        ans = sign ? ans : -1 * ans;

        return (int)ans;
    }
};

int main()
{
    Solution a;
    string s;
    s = "42";
    std::cout << a.myAtoi(s) << std::endl;
    s = "-91283472332";
    std::cout << a.myAtoi(s) << std::endl;
    s = "20000000000000000000";
    std::cout << a.myAtoi(s) << std::endl;
    s = "+4193";
    std::cout << a.myAtoi(s) << std::endl;
    s = "41";
    std::cout << a.myAtoi(s) << std::endl;
    s = "          -42";
    std::cout << a.myAtoi(s) << std::endl;
    s = "43 with many words";
    std::cout << a.myAtoi(s) << std::endl;
    s = "many words but i see 0";
    std::cout << a.myAtoi(s) << std::endl;
    s = "4193";
    std::cout << a.myAtoi(s) << std::endl;
}
